File: testing/iOS/StyleConverterTest/StyleEngine/layout/LayoutExtractor.swift
//
//  LayoutExtractor.swift
//  StyleEngine/layout — Phase 7, step 1 (scaffold only).
//
//  Facade over every layout triplet extractor. Mirrors `TypographyExtractor`
//  (Phase 6) in shape: one `extract(from:)` entry point returning a
//  populated `LayoutAggregate?`. For THIS step the function returns `nil`
//  unconditionally — no per-triplet extractors exist yet. Steps 2-5 fill
//  in the flexbox, grid, position, advanced, and root groups respectively.
//
//  What this file *does* do today: register the 60 layout property type
//  names with `PropertyRegistry.migrated` so the renderer knows those
//  properties are owned by the layout engine (even if the engine currently
//  no-ops them — the legacy StyleBuilder still handles them in practice
//  because `LayoutApplier.apply(...)` is identity and callers read the
//  migration ledger via `union`-ing the set below).
//
//  Important: the 60 PascalCase type names below mirror the IR type
//  strings produced by `src/main/kotlin/app/parsing/css/properties/
//  longhands/PropertyParserRegistry.kt` (category folders:
//  layout/flexbox, layout/grid, layout/position, layout/advanced, plus
//  the four "root" longhands). If a name fails to parse on the CSS side,
//  the registration here is harmless — it just flags a non-existent
//  property as owned.
//

import Foundation

// MARK: - Property-name groupings (public so PropertyRegistry can union them)

/// Flexbox family — 12 longhands under layout/flexbox/.
/// Source of truth: irmodels/properties/layout/flexbox/ and the matching
/// parser folder.
enum LayoutFlexboxProperty {
    /// Exhaustive name set for Phase 7 step 2 (flexbox).
    static let set: Set<String> = [
        // Container-level flex keywords.
        "Display", "FlexDirection", "FlexWrap",
        // Per-item flex sizing.
        "FlexGrow", "FlexShrink", "FlexBasis",
        // Main/cross-axis alignment.
        "JustifyContent", "AlignItems", "AlignContent", "AlignSelf",
        // Child ordering.
        "Order",
        // Legacy flex-box keyword retained in the IR for fidelity.
        "BoxOrient",
    ]
}

/// Grid family — 18 longhands under layout/grid/.
enum LayoutGridProperty {
    /// Exhaustive name set for Phase 7 step 3 (grid).
    static let set: Set<String> = [
        // Explicit track templates.
        "GridTemplateColumns", "GridTemplateRows", "GridTemplateAreas",
        // Implicit (auto) tracks.
        "GridAutoColumns", "GridAutoRows", "GridAutoFlow",
        // IR also carries a unified GridAutoTrack for parser convenience.
        "GridAutoTrack",
        // Item placement.
        "GridArea",
        "GridColumnStart", "GridColumnEnd",
        "GridRowStart",    "GridRowEnd",
        // Grid-level alignment (justify-items / justify-self are grid-only
        // in practice — kept here so they co-locate with track setup).
        "JustifyItems", "JustifySelf",
        // Multi-track alignment (rare; spec-defined, rarely implemented).
        "AlignTracks", "JustifyTracks",
        // Masonry (CSS Grid Level 3 draft).
        "MasonryAutoFlow",
        // `grid-template` shorthand's IR form.
        "GridTemplate",
    ]
}

/// Position family — 10 longhands under layout/position/.
enum LayoutPositionProperty {
    /// Exhaustive name set for Phase 7 step 4 (position + inset).
    static let set: Set<String> = [
        // Core scheme selector.
        "Position",
        // Physical inset longhands.
        "Top", "Right", "Bottom", "Left",
        // Logical inset longhands (resolved against writing-mode later).
        "InsetBlockStart", "InsetBlockEnd",
        "InsetInlineStart", "InsetInlineEnd",
        // Paint-order.
        "ZIndex",
    ]
}

/// Advanced positioning / motion-path family — 17 longhands under
/// layout/advanced/. CSS Anchor Positioning + Motion Path modules.
enum LayoutAdvancedProperty {
    /// Exhaustive name set for Phase 7 step 5 (advanced). SwiftUI has no
    /// direct analogue for most of these — the eventual applier will log
    /// TODOs, but registering the names here keeps coverage honest.
    static let set: Set<String> = [
        // CSS Anchor Positioning.
        "AnchorName", "AnchorScope", "InsetArea",
        // CSS Motion Path.
        "OffsetPath", "OffsetDistance", "OffsetAnchor",
        "OffsetPosition", "OffsetRotate",
        // `offset` shorthand IR form.
        "Offset",
        // Anchor Positioning fallback chain.
        "PositionAnchor", "PositionArea",
        "PositionFallback",
        "PositionTry", "PositionTryFallbacks",
        "PositionTryOptions", "PositionTryOrder",
        "PositionVisibility",
    ]
}

/// Root / flow family — 4 top-level longhands that don't fit flex/grid/
/// position. `Overlay` + `ReadingFlow` are CSS Display Level 4 additions.
enum LayoutRootProperty {
    /// Exhaustive name set for Phase 7 step 5 (root).
    static let set: Set<String> = [
        // Classic float model — rarely used in SDUI content.
        "Clear", "Float",
        // CSS Display Level 4 — flow overlay.
        "Overlay",
        // Accessibility-driven reading-order override.
        "ReadingFlow",
    ]
}

/// Union of all five layout groups — 60 property type names total.
/// `PropertyRegistry.migrated` picks this up via `.union(...)`.
enum LayoutProperty {
    /// All layout-family property type names. The count is asserted by
    /// `LayoutSelfTest.run()` to catch accidental drift.
    static let set: Set<String> =
        LayoutFlexboxProperty.set
            .union(LayoutGridProperty.set)
            .union(LayoutPositionProperty.set)
            .union(LayoutAdvancedProperty.set)
            .union(LayoutRootProperty.set)
}

// MARK: - Extractor facade

enum LayoutExtractor {

    /// Runs every (future) layout extractor and folds them into one
    /// aggregate. Returns nil when no layout property was seen — this
    /// lets `LayoutApplier` short-circuit.
    ///
    /// STEP 1 IMPLEMENTATION: no extractors exist yet, so this always
    /// returns nil. Steps 2-5 will replace this with the typography-style
    /// `Applier.contribute(Extractor.extract(from:), into: &agg)` chain.
    /// Keeping the signature stable now means downstream callers
    /// (StyleBuilder / ComponentRenderer) can be wired in step 2 without
    /// churn.
    static func extract(from properties: [IRProperty]) -> LayoutAggregate? {
        // Start with an empty aggregate; each sub-extractor folds in
        // its own fields and flips `touched` when it writes.
        var agg = LayoutAggregate()

        // Step 2 — flexbox family. 11 properties are read
        // here; BoxOrient is intentionally skipped per task brief.
        FlexboxExtractor.extract(from: properties, into: &agg)

        // Step 3 — grid family (GridTemplate*/GridAuto*/GridArea +
        // JustifyItems/JustifySelf). Writes tracks/areas/flow/placement
        // into the aggregate's grid-* slots.
        GridExtractor.contribute(properties, into: &agg)

        // Step 4 — position family (Position/Top/Right/Bottom/Left +
        // InsetBlock*/InsetInline* + ZIndex). Resolves logical insets
        // to physical sides against the default (LTR) layout direction.
        PositionExtractor.contribute(properties, into: &agg)

        // Steps 5 (advanced / root) plug in next to this call —
        // parallel agents own those scopes.

        // Return nil when nothing wrote — lets LayoutApplier short-
        // circuit and leaves the renderer's legacy fallback untouched.
        return agg.touched ? agg : nil
    }
}

File: testing/Android/app/src/main/java/com/styleconverter/test/style/layout/LayoutExtractor.kt
package com.styleconverter.test.style.layout

import com.styleconverter.test.style.PropertyRegistry
import kotlinx.serialization.json.JsonElement

/**
 * Phase 7 step 1 scaffold extractor for the CSS layout family.
 *
 * Responsibilities:
 *   1. Claim every layout IR property name against [PropertyRegistry] so the
 *      legacy [com.styleconverter.test.style.StyleApplier] dispatch switch
 *      defers to the style-engine path instead of double-applying.
 *   2. Expose a single [extractLayoutConfig] entrypoint that returns a fully
 *      resolved [LayoutConfig]. In step 1 every field resolves to null — the
 *      real extraction logic arrives in Phase 7 steps 2–5.
 *
 * Precedent: [com.styleconverter.test.style.typography.TypographyExtractor] —
 * same init-block registration + single `extract*Config` entrypoint shape.
 *
 * Owner strings match the IR sub-folder name for introspection-friendly
 * coverage audits via [PropertyRegistry.allRegistered]. The four sub-folders
 * are: layout/flexbox, layout/grid, layout/position, layout/advanced; plus
 * the four layout root properties (Clear / Float / Overlay / ReadingFlow)
 * registered under "layout".
 */
object LayoutExtractor {

    init {
        // ----- layout/flexbox -----
        // See src/main/kotlin/app/irmodels/properties/layout/flexbox/*.kt for
        // the source-of-truth property files. Each name below maps 1:1 to an
        // IRProperty.type emitted by the CSS parser.
        PropertyRegistry.migrated(
            "Display",
            "FlexDirection", "FlexWrap",
            "FlexGrow", "FlexShrink", "FlexBasis",
            "JustifyContent", "AlignItems", "AlignContent",
            "AlignSelf", "Order",
            // Legacy box-orient lives under flexbox in the IR tree because it
            // predates the flex spec but shares axis semantics.
            "BoxOrient",
            owner = "layout/flexbox"
        )

        // ----- layout/grid -----
        // Mirrors src/main/kotlin/app/irmodels/properties/layout/grid/*.kt.
        // GridTemplate (shorthand) + GridAutoTrack (CSS Grid 3 proposal) are
        // parse-only at this stage but the tripwire test requires them listed.
        PropertyRegistry.migrated(
            "GridTemplateColumns", "GridTemplateRows", "GridTemplateAreas",
            "GridTemplate",
            "GridAutoColumns", "GridAutoRows", "GridAutoFlow",
            "GridAutoTrack",
            "GridArea",
            "GridColumnStart", "GridColumnEnd",
            "GridRowStart", "GridRowEnd",
            "JustifyItems", "JustifySelf",
            "AlignTracks", "JustifyTracks",
            "MasonryAutoFlow",
            owner = "layout/grid"
        )

        // ----- layout/position -----
        // Mirrors src/main/kotlin/app/irmodels/properties/layout/position/.
        // Logical insets (InsetBlock*, InsetInline*) register alongside
        // physical top/right/bottom/left; step 4 will fold them into a
        // single InsetRect via LayoutDirection.
        PropertyRegistry.migrated(
            "Position",
            "Top", "Right", "Bottom", "Left",
            "InsetBlockStart", "InsetBlockEnd",
            "InsetInlineStart", "InsetInlineEnd",
            "ZIndex",
            owner = "layout/position"
        )

        // ----- layout/advanced -----
        // CSS Anchor Positioning (2024) + CSS Motion Path + Position Try.
        // All parse-only on Compose today — registering so the legacy switch
        // doesn't silently claim them while the style-engine catches up.
        PropertyRegistry.migrated(
            "AnchorName", "AnchorScope",
            "InsetArea",
            "OffsetPath", "OffsetDistance", "OffsetAnchor", "OffsetPosition",
            "OffsetRotate", "Offset",
            "PositionAnchor", "PositionArea",
            "PositionFallback", "PositionTry",
            "PositionTryFallbacks", "PositionTryOptions",
            "PositionTryOrder",
            "PositionVisibility",
            owner = "layout/advanced"
        )

        // ----- layout root -----
        // Four top-level IR files directly under layout/: Clear / Float /
        // Overlay / ReadingFlow. FloatExtractor already handles Clear+Float
        // at runtime but did not previously register with PropertyRegistry —
        // we claim them here so the tripwire test stays green.
        PropertyRegistry.migrated(
            "Clear", "Float",
            "Overlay",
            "ReadingFlow",
            owner = "layout"
        )
    }

    /**
     * Extract a fully-resolved [LayoutConfig] from the IR property stream.
     *
     * Step 1 stub: returns [LayoutConfig.Empty] unconditionally. Subsequent
     * Phase 7 steps replace the stub with real per-property extraction. The
     * method signature is locked in now so the facade and ComponentRenderer
     * hook don't churn between steps.
     *
     * @param properties (propertyType, data) pairs from IRComponent.properties.
     * @return LayoutConfig with all fields null in step 1.
     */
    fun extractLayoutConfig(properties: List<Pair<String, JsonElement?>>): LayoutConfig {
        // Phase 7b step 2: delegate to the flexbox sub-extractor. The flexbox
        // extractor returns a LayoutConfig with ONLY its owned fields
        // populated; all grid/position/advanced/root slots stay null. When
        // later steps ship their own sub-extractors (GridExtractor,
        // PositionExtractor) they'll merge into this same aggregate via a
        // copy-based fold — see the TODO below.
        val flexbox = com.styleconverter.test.style.layout.flexbox.FlexboxExtractor
            .extract(properties)

        // Phase 7b — fold grid + position sub-extractors into the flexbox
        // result. Each sub-extractor returns only its own fields populated;
        // we merge by copying non-null fields onto the aggregate. The
        // advanced/root extractors will join via the same pattern in a
        // later step.
        val g = com.styleconverter.test.style.layout.grid.GridLayoutExtractor.extract(properties)
        val p = com.styleconverter.test.style.layout.position.PositionLayoutExtractor.extract(properties)
        return flexbox.copy(
            gridTemplateColumns = g.templateColumns ?: flexbox.gridTemplateColumns,
            gridTemplateRows = g.templateRows ?: flexbox.gridTemplateRows,
            gridTemplateAreas = g.templateAreas ?: flexbox.gridTemplateAreas,
            gridAutoColumns = g.autoColumns ?: flexbox.gridAutoColumns,
            gridAutoRows = g.autoRows ?: flexbox.gridAutoRows,
            gridAutoFlow = g.autoFlow ?: flexbox.gridAutoFlow,
            gridArea = g.gridArea ?: flexbox.gridArea,
            gridColumn = g.gridColumn ?: flexbox.gridColumn,
            gridRow = g.gridRow ?: flexbox.gridRow,
            // justifyItems/justifySelf are shared with flexbox's JustifyContent
            // alignment set; if flexbox already claimed them prefer that so a
            // container that happens to set both gets the flex semantics. In
            // practice only one code path populates each per component.
            justifyItems = flexbox.justifyItems ?: g.justifyItems,
            justifySelf = flexbox.justifySelf ?: g.justifySelf,
            position = p.position ?: flexbox.position,
            inset = p.inset ?: flexbox.inset,
            zIndex = p.zIndex ?: flexbox.zIndex,
        )
    }
}
